dbms/src/DataTypes/DataTypeFactory.h

```diff
@@ -38,10 +38,14 @@ class DataTypeFactory final : public ext::Singleton<DataTypeFactory>
 private:
     using Creator = std::function<DataTypePtr(const ASTPtr & parameters)>;
     using SimpleCreator = std::function<DataTypePtr()>;
+    // family_name -> Creator
     using DataTypesDictionary = std::unordered_map<String, Creator>;
+    // full_name -> DataTypePtr
+    using FullnameTypes = std::unordered_map<String, DataTypePtr>;
 
 public:
     DataTypePtr get(const String & full_name) const;
+    DataTypePtr getOrSet(const String & full_name);
```

Comment:
```suggestion
    // In order to optimize the speed of generating data type instances, this will cache
    // the full_name -> DataTypePtr
    DataTypePtr getOrSet(const String & full_name);
```